### PR TITLE
Monitor security analysis

### PR DESCRIPTION
**Pull request by EtienneLt**
Signed-off-by: Etienne Lesot <etienne.lesot@rte-france.com>

**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
#1699

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*

**What is the current behavior?** *(You can also link to an open issue here)*

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysis.java [hunk 1]
```diff
@@ -225,10 +252,18 @@ private SecurityAnalysisResult runSimulations(LfNetwork network, List<Propagated
...
         // create a contingency list that impact the network
         List<LfContingency> contingencies = createContingencies(propagatedContingencies, network);
+        List<BranchResult> preContingencyBranchResults = new ArrayList<>();
+        List<BusResults> preContingencyBusResults = new ArrayList<>();
 
         // run pre-contingency simulation
         try (AcloadFlowEngine engine = new AcloadFlowEngine(network, acParameters)) {
             AcLoadFlowResult preContingencyLoadFlowResult = engine.run(Reporter.NO_OP);
+            monitors.values().forEach(stateMonitor -> {
+                if (stateMonitor.getContingencyContext().getContextType() == ContingencyContextType.ALL ||
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysis.java:262]: @sylvlecl Reading again this, I am not sure that our choice on `powsybl-core` side is the good one...

File: src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysis.java [hunk 2]
```diff
@@ -302,6 +340,10 @@ private PostContingencyResult runPostContingencySimulation(LfNetwork network, Ac
                 network.getBranches().stream().filter(b -> !lfContingency.getBranches().contains(b)),
                 network.getBuses().stream().filter(b -> !lfContingency.getBuses().contains(b)),
                 postContingencyLimitViolations);
+            StateMonitor stateMonitor = monitors.get(lfContingency.getContingency().getId());
+            if (stateMonitor != null && stateMonitor.getContingencyContext().getContextType() == ContingencyContextType.SPECIFIC) {
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysis.java:344]: Maybe you have to implement the `ALL` mode in a different way to be more efficient, but I am not sure about that.

File: src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysis.java [hunk 3]
```diff
@@ -342,4 +386,22 @@ private static boolean violationWeakenedOrEquivalent(LimitViolation violation1,
...
+
+    private void addMonitorInfos(LfNetwork lfNetwork, StateMonitor monitor, Consumer<BranchResult> branchResultConsumer,
+                                 Consumer<BusResults> busResultsConsumer) {
+        monitor.getBranchIds().forEach(branchId -> branchResultConsumer.accept(createBranchResult(lfNetwork.getBranchById(branchId))));
+        lfNetwork.getBuses().stream().filter(lfBus -> monitor.getVoltageLevelIds().contains(lfBus.getVoltageLevelId())).forEach(lfBus ->
+            busResultsConsumer.accept(createBusResult(lfBus, lfBus.getVoltageLevelId())));
+    }
+
+    private BranchResult createBranchResult(LfBranch branch) {
+        double currentScale = PerUnit.SB / branch.getBus1().getNominalV();
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysis.java:398]: Here I am not sure that it is robust to get the nominal voltage through the bus. I think you should try to put this method in the LfBranch directly and use attributes.

File: src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysis.java [hunk 4]
```diff
@@ -342,4 +386,22 @@ private static boolean violationWeakenedOrEquivalent(LimitViolation violation1,
...
+        return new BranchResult(branch.getId(), branch.getP1().eval() * PerUnit.SB, branch.getQ1().eval(), branch.getI1().eval() * currentScale,
+            branch.getP2().eval() * PerUnit.SB, branch.getQ2().eval() * PerUnit.SB, branch.getI2().eval() * currentScale);
+    }
+
+    private BusResults createBusResult(LfBus bus, String voltageLevelId) {
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysis.java:403]: Same here: please try to put that inside the LfBus.

**Review by annetill — CHANGES_REQUESTED** [review]
Some changes are needed but globally it is correct.

File: src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysis.java (hunk 3 above)
Comment (annetill) [reply src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysis.java:398]: Add a unit test with open branches, dangline line and three windings transformer.

File: src/main/java/com/powsybl/openloadflow/network/impl/LfNetworkLoaderImpl.java [hunk 5]
```diff
@@ -259,9 +259,13 @@ private static void createBranches(List<LfBus> lfBuses, LfNetwork lfNetwork, Loa
             LfBus lfBus1 = getLfBus(t3wt.getLeg1().getTerminal(), lfNetwork, parameters.isBreakers());
             LfBus lfBus2 = getLfBus(t3wt.getLeg2().getTerminal(), lfNetwork, parameters.isBreakers());
             LfBus lfBus3 = getLfBus(t3wt.getLeg3().getTerminal(), lfNetwork, parameters.isBreakers());
-            addBranch(lfNetwork, LfLegBranch.create(lfNetwork, lfBus1, lfBus0, t3wt, t3wt.getLeg1(), parameters.isTwtSplitShuntAdmittance()), report);
-            addBranch(lfNetwork, LfLegBranch.create(lfNetwork, lfBus2, lfBus0, t3wt, t3wt.getLeg2(), parameters.isTwtSplitShuntAdmittance()), report);
-            addBranch(lfNetwork, LfLegBranch.create(lfNetwork, lfBus3, lfBus0, t3wt, t3wt.getLeg3(), parameters.isTwtSplitShuntAdmittance()), report);
+            LfLegBranch lfBranch1 = LfLegBranch.create(lfNetwork, lfBus1, lfBus0, t3wt, t3wt.getLeg1(), parameters.isTwtSplitShuntAdmittance());
```
Comment (sylvlecl) [inline src/main/java/com/powsybl/openloadflow/network/impl/LfNetworkLoaderImpl.java:262]: The last argument (report) has been removed in those calls : I think it's not what you wanted to do !

File: src/main/java/com/powsybl/openloadflow/network/AbstractLfBranch.java [hunk 6]
```diff
@@ -205,4 +206,22 @@ public boolean isVoltageController() {
...
     }
+
+    @Override
+    public BranchResult createBranchResult() {
+        if (getBus1() == null && getBus2() == null) {
+            return new BranchResult(getId(), 0, 0, 0, 0, 0, 0);
+        } else if (getBus1() == null) {
+            double currentScale = PerUnit.SB / getBus2().getNominalV();
+            return new BranchResult(getId(), 0, 0, 0,
+                getP2().eval() * PerUnit.SB, getQ2().eval() * PerUnit.SB, getI2().eval() * currentScale);
```
Comment (sylvlecl) [inline src/main/java/com/powsybl/openloadflow/network/AbstractLfBranch.java:217]: Probably to do in another PR, but I think it would make sense to share the scaling methods in the `PerUnit` class, what do you think @annetill @geofjamg ?
Like
```java
class PerUnit {
    double scaleP(double p) { ... }
    double unscaleP(double p)  { ... }
    double scaleI(double i, double nominalV)  { ... }
    double unscaleI(double i, double nominalV)  { ... }
    ...
```

It would reduce a little the boiler plate code and possibilities for mistakes.

File: src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysis.java [hunk 7]
```diff
@@ -60,15 +68,32 @@
...
+    public OpenSecurityAnalysis(Network network, LimitViolationDetector detector, LimitViolationFilter filter,
+                                MatrixFactory matrixFactory, Supplier<GraphDecrementalConnectivity<LfBus>> connectivityProvider, List<StateMonitor> stateMonitors) {
         this.network = Objects.requireNonNull(network);
         this.detector = Objects.requireNonNull(detector);
         this.filter = Objects.requireNonNull(filter);
         this.matrixFactory = Objects.requireNonNull(matrixFactory);
         this.connectivityProvider = Objects.requireNonNull(connectivityProvider);
+        Objects.requireNonNull(stateMonitors);
+        this.monitors = new HashMap<>();
+        stateMonitors.forEach(monitor -> {
```
Comment (sylvlecl) [inline src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysis.java:89]: let's not forget to re-use what you have done in core once it's done :)

File: src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysis.java (hunk 1 above)
Comment (sylvlecl) [reply src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysis.java:264]: You mean to include the "ALL" into the N situation ?

File: src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysis.java [hunk 8]
```diff
@@ -269,7 +307,11 @@ private SecurityAnalysisResult runSimulations(LfNetwork network, List<Propagated
             }
 
             LimitViolationsResult preContingencyResult = new LimitViolationsResult(preContingencyComputationOk, new ArrayList<>(preContingencyLimitViolations.values()));
-            return new SecurityAnalysisResult(preContingencyResult, postContingencyResults);
+            return new SecurityAnalysisResult(preContingencyResult, postContingencyResults,
```
Comment (sylvlecl) [inline src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysis.java:310]: Not a must-do, but maybe it would be easier to do building of the maps inside the constructor, so that we do not have to do it ourselves here

File: src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysis.java [hunk 9]
```diff
@@ -312,16 +361,20 @@ private PostContingencyResult runPostContingencySimulation(LfNetwork network, Ac
...
 
         LfContingency.reactivateEquations(deactivatedEquations, deactivatedEquationTerms);
-
         stopwatch.stop();
         LOGGER.info("Post contingency '{}' simulation done in {} ms", lfContingency.getContingency().getId(),
                 stopwatch.elapsed(TimeUnit.MILLISECONDS));
 
-        return new PostContingencyResult(lfContingency.getContingency(), postContingencyComputationOk, new ArrayList<>(postContingencyLimitViolations.values()));
+        return new PostContingencyResult(lfContingency.getContingency(), postContingencyComputationOk, new ArrayList<>(postContingencyLimitViolations.values()),
+            branchResults.stream().collect(Collectors.toMap(BranchResult::getBranchId, Function.identity())),
```
Comment (sylvlecl) [inline src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysis.java:369]: Same here, maybe it would be simpler to change the constructor so that we just provide the lists

File: src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysis.java [hunk 10]
```diff
@@ -342,4 +395,13 @@ private static boolean violationWeakenedOrEquivalent(LimitViolation violation1,
     List<LfContingency> createContingencies(List<PropagatedContingency> propagatedContingencies, LfNetwork network) {
         return LfContingency.createContingencies(propagatedContingencies, network, network.createDecrementalConnectivity(connectivityProvider), true);
     }
+
+    private void addMonitorInfos(LfNetwork lfNetwork, StateMonitor monitor, Consumer<BranchResult> branchResultConsumer,
+                                 Consumer<BusResults> busResultsConsumer, Consumer<ThreeWindingsTransformerResult> threeWindingsTransformerResultConsumer) {
+        monitor.getBranchIds().forEach(branchId -> branchResultConsumer.accept(lfNetwork.getBranchById(branchId).createBranchResult()));
```
Comment (sylvlecl) [inline src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysis.java:401]: I wonder what we should do in case of an invalid branch ID (one that does not existe in the network.

I guess that with this code, it will throw a null pointer exception. I would be in favor of silently ignore that branch instead.
In any case we must add a unit test to cover that case! (same for buses and T3)

File: src/main/java/com/powsybl/openloadflow/network/impl/LfThreeWindingsTransformer.java [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+    public LfThreeWindingsTransformer(String lfThreeWindingsTransformerId, LfLegBranch branch1, LfLegBranch branch2, LfLegBranch branch3) {
+        this.lfThreeWindingsTransformerId = Objects.requireNonNull(lfThreeWindingsTransformerId);
+        this.branch1 = Objects.requireNonNull(branch1);
+        this.branch2 = Objects.requireNonNull(branch2);
+        this.branch3 = Objects.requireNonNull(branch3);
+    }
+
+    public ThreeWindingsTransformerResult createThreeWindingsTransformerResult() {
+        return new ThreeWindingsTransformerResult(lfThreeWindingsTransformerId, branch1.getP1().eval(), branch1.getQ1().eval(),
```
Comment (sylvlecl) [inline src/main/java/com/powsybl/openloadflow/network/impl/LfThreeWindingsTransformer.java:31]: we probably need to handle the disconnected cases here, like for the branches (checking that buses are not null) : we'll need a unit test for this too.

File: src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysis.java [hunk 12]
```diff
@@ -302,6 +346,11 @@ private PostContingencyResult runPostContingencySimulation(LfNetwork network, Ac
                 network.getBranches().stream().filter(b -> !lfContingency.getBranches().contains(b)),
                 network.getBuses().stream().filter(b -> !lfContingency.getBuses().contains(b)),
                 postContingencyLimitViolations);
+            StateMonitor stateMonitor = monitors.get(lfContingency.getContingency().getId());
+            if (stateMonitor != null && (stateMonitor.getContingencyContext().getContextType() == ContingencyContextType.ALL
+                || stateMonitor.getContingencyContext().getContextType() == ContingencyContextType.SPECIFIC)) {
+                addMonitorInfos(network, stateMonitor, branchResults::add, busResults::add, threeWindingsTransformerResults::add);
```
Comment (sylvlecl) [inline src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysis.java:352]: Maybe there is something to do here to filter out buses and branches which are disconnected by the contingency, see filters above in the identification of violations.

Not 100% sure about that : @annetill and @geofjamg do you know if there is already a test case covering that ?

File: src/main/java/com/powsybl/openloadflow/network/impl/LfNetworkLoaderImpl.java (hunk 5 above)
Comment (EtienneLt) [reply src/main/java/com/powsybl/openloadflow/network/impl/LfNetworkLoaderImpl.java:262]: look at the lines below report is there

Comment (sylvlecl) [reply src/main/java/com/powsybl/openloadflow/network/impl/LfNetworkLoaderImpl.java:262]: you're right, my mistake!

File: src/main/java/com/powsybl/openloadflow/network/AbstractLfBranch.java (hunk 6 above)
Comment (annetill) [reply src/main/java/com/powsybl/openloadflow/network/AbstractLfBranch.java:217]: Yes, great idea, per unit system must be hidden. I will make it myself I think.

File: src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysis.java (hunk 12 above)
Comment (annetill) [reply src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysis.java]: yes, use `isDisabled` for buses. For branches, the method has to be created.

File: src/main/java/com/powsybl/openloadflow/network/AbstractLfBranch.java (hunk 6 above)
Comment (sylvlecl) [reply src/main/java/com/powsybl/openloadflow/network/AbstractLfBranch.java:217]: An afterthought : actually it could even be an interface, that way we could use different per-unit systems and compare their effect  !! :)  But I admit, it's not really useful in practice ...

File: src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysis.java [hunk 13]
```diff
@@ -225,10 +245,17 @@ private SecurityAnalysisResult runSimulations(LfNetwork network, List<Propagated
...
         // create a contingency list that impact the network
         List<LfContingency> contingencies = createContingencies(propagatedContingencies, network);
+        List<BranchResult> preContingencyBranchResults = new ArrayList<>();
+        List<BusResults> preContingencyBusResults = new ArrayList<>();
+        List<ThreeWindingsTransformerResult> preContingencyThreeWindingsTransformerResults = new ArrayList<>();
 
         // run pre-contingency simulation
         try (AcloadFlowEngine engine = new AcloadFlowEngine(network, acParameters)) {
             AcLoadFlowResult preContingencyLoadFlowResult = engine.run(Reporter.NO_OP);
+            addMonitorInfos(network, monitorIndex.getNoneStateMonitor(), preContingencyBranchResults::add, preContingencyBusResults::add,
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysis.java:255]: I understand why it works but it could be more efficient to check the contingency context here, no?

**Review by geofjamg — APPROVED** [review]

**Review by annetill — APPROVED** [review]
